internal/controllers/group_messages_controller.go

```diff
@@ -0,0 +1,135 @@
+package controllers
+
+import (
+	"github.com/gin-gonic/gin"
+	"net/http"
+	"strconv"
+
+	"github.com/google/uuid"
+
+	"Groupchat-Service/internal/models"
+	"Groupchat-Service/internal/services"
+)
+
+// FCMMessageController handles HTTP requests for messages
+type FCMMessageController struct {
+	messageService services.MessageService
+}
+
+func NewMessageController(messageService services.MessageService) *FCMMessageController {
+	return &FCMMessageController{
+		messageService: messageService,
+	}
+}
+
+// GetMessages handles GET /groups/messages
+// GetMessages handles GET /groups/messages
+func (c *FCMMessageController) GetMessages(ctx *gin.Context) {
+	// Extract group ID from URL
+	groupID, err := uuid.Parse(ctx.Param("groupId"))
+	if err != nil {
+		ctx.JSON(http.StatusBadRequest, gin.H{"error": "Invalid group ID"})
+		return
+	}
+
+	// Parse pagination query
+	query := models.PaginationQuery{
+		PageSize:  10, // Default page size
+		Direction: "next",
+	}
+
+	if size := ctx.Query("pageSize"); size != "" {
+		pageSize, err := strconv.Atoi(size)
+		if err != nil || pageSize < 1 || pageSize > 50 {
+			ctx.JSON(http.StatusBadRequest, gin.H{"error": "page size must be between 1 and 50"})
+			return
+		}
+		query.PageSize = pageSize
+	}
+
+	if cursor := ctx.Query("cursor"); cursor != "" {
+		query.Cursor = &cursor
+	}
+
+	if direction := ctx.Query("direction"); direction != "" {
+		if direction != "next" && direction != "previous" {
+			ctx.JSON(http.StatusBadRequest, gin.H{"error": "direction must be 'next' or 'previous'"})
+			return
+		}
+		query.Direction = direction
+	}
+
+	if search := ctx.Query("search"); search != "" {
+		if len(search) > 100 {
+			ctx.JSON(http.StatusBadRequest, gin.H{"error": "search term too long: maximum 100 characters"})
+			return
+		}
+		query.Search = &search
+	}
+
```

Comment:
all these checks make your controller bloated, and seeing as how it's business logic, i think its best to move this to either a service layer, or a util method that would check the entire response.
avoid magic numbers. Put 100 and 50 in constants. They're fine for things like this as the context is explained just before:
``` PageSize: 10 ```

Reply:
From what I know the checks/validation isn't seen as buisiness logic, but I'll seperate it for you ;)